src/client/index.ts

```diff
@@ -28,30 +35,157 @@ class ErrorHandler {
+  // root. For now we'll just run LSP per workspace.
+  const root = vscode.workspace.rootPath;
+  if (root == null) {
+    return false;
+  }
+
+  const esyJson = path.join(root, "esy.json");
+  const packageJson = path.join(root, "package.json");
+  if (await exists(esyJson)) {
+    return true;
+  } else if (await exists(packageJson)) {
+    // package.json could be unrelated to esy, check if it has "esy" config
+    // then.
+    try {
+      const data = await readFile(packageJson, "utf8");
+      const json = JSON.parse(data);
+      return json.esy != null;
+    } catch (_e) {
+      return false;
+    }
+  }
+
+  return false;
+}
+
+async function isBSProject() {
+  // TODO: we need to use workspace.workspaceFolders here and run LSP server per
+  // root. For now we'll just run LSP per workspace.
+  const root = vscode.workspace.rootPath;
+  if (root == null) {
+    return false;
+  }
+
+  const bsconfigJson = path.join(root, "bsconfig.json");
+
+  if (await exists(bsconfigJson)) {
+    return true;
+  }
+
+  return false;
+}
+
 export async function launch(context: vscode.ExtensionContext): Promise<void> {
+  const isEasyProject = await isEsyProject();
+  const isBucklescriptProject = await isBSProject();
+  return launchMerlinLsp(context, {
+    isBucklescriptProject,
+    useEsy: isEasyProject,
+  });
+}
+
+function getPrebuiltExecutablesPath() {
+  return path.join(__dirname, `../../../executables/${process.platform}`);
+}
+
+function getMerlinLspPath(isBucklescriptProject: boolean) {
   const reasonConfig = vscode.workspace.getConfiguration("reason");
-  const module = context.asAbsolutePath(path.join("node_modules", "ocaml-language-server", "bin", "server"));
-  const options = { execArgv: ["--nolazy", "--inspect=6009"] };
-  const transport = client.TransportKind.ipc;
-  const run = { module, transport };
-  const debug = {
-    module,
-    options,
-    transport,
+  let merlinLspPath = reasonConfig.get<string | null>("path.ocamlmerlin-lsp", null);
+
+  if (merlinLspPath == null) {
+    merlinLspPath = isWin ? "ocamlmerlin-lsp.exe" : "ocamlmerlin-lsp";
+
+    if (isBucklescriptProject) {
+      merlinLspPath = path.join(getPrebuiltExecutablesPath(), merlinLspPath);
+    }
+  }
+
+  return merlinLspPath;
+}
+
+function getMerlinReasonDir() {
+  const reasonConfig = vscode.workspace.getConfiguration("reason");
+  const merlinReasonPath = reasonConfig.get<string | null>("path.ocamlmerlin-reason", null);
+
+  if (merlinReasonPath == null) return getPrebuiltExecutablesPath();
+
+  return path.dirname(merlinReasonPath);
+}
+
+function getMerlinLspOptions(options: { useEsy: boolean; isBucklescriptProject: boolean }) {
+  const merlinLsp = getMerlinLspPath(options.isBucklescriptProject);
+  const merlinReasonDir = getMerlinReasonDir();
+
+  let run;
+  if (options.useEsy) {
+    run = {
+      args: ["exec-command", "--include-current-env", merlinLsp],
+      command: process.platform === "win32" ? "esy.cmd" : "esy",
```

Comment:
The way this is implemented has some unexpected behavior. If I'm understanding correctly, if user sets in `settings.json`:

```json
{
  "reason.path.ocamlmerlin-lsp": "/some/folder/ocamlmerlin-lsp"
}
```

then the command will be
```bash
esy /some/folder/ocamlmerlin-lsp
```

but I would expect it to be just 

```bash
/some/folder/ocamlmerlin-lsp
```

If the path is set manually, then it should be used "as is", imo.